cmd/snap/cmd_model.go

```diff
@@ -0,0 +1,187 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+	"gopkg.in/yaml.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+)
+
+var (
+	shortModelHelp = i18n.G("Get the active model for this device")
+	longModelHelp  = i18n.G(`
+The model command returns only the active model assertion information for this 
+device.	
+
+The --serial option allows printing the serial assertion instead of the model
+assertion.
+
+The --verbose option prints off more headers from the assertion.
+
+The --assertion option prints off the raw assertion.
+`)
+)
+
+type cmdModel struct {
+	waitMixin
+	Serial    bool `long:"serial"`
+	Verbose   bool `long:"verbose"`
+	Assertion bool `long:"assertion"`
+}
+
+func init() {
+	cmd := addCommand("model",
+		shortModelHelp,
+		longModelHelp,
+		func() flags.Commander {
+			return &cmdModel{}
+		}, waitDescs.also(map[string]string{
+			"assertion": i18n.G("Print the raw assertion."),
+			"verbose":   i18n.G("Print more of the assertion fields."),
+			"serial":    i18n.G("Print the serial assertion instead of the model."),
+		}),
+		[]argDesc{},
+	)
+	cmd.hidden = true
+}
+
+func (x *cmdModel) Execute(args []string) error {
+	var assertion asserts.Assertion
+	var err error
+	if x.Serial {
+		assertion, err = x.client.CurrentSerialAssertion()
+	} else {
+		assertion, err = x.client.CurrentModelAssertion()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// --assertion means output the raw assertion
+	if x.Assertion {
+		_, err = Stdout.Write(asserts.Encode(assertion))
+		return err
+	}
+
+	termWidth, _ := termSize()
+	termWidth -= 3
+	if termWidth > 100 {
+		// any wider than this and it gets hard to read
+		termWidth = 100
+	}
+
+	w := tabWriter()
+
+	primaryHeaderKeys := assertion.Type().PrimaryKey
+	for _, header := range primaryHeaderKeys {
+		val := assertion.HeaderString(header)
+		err = maybeWrapSingleYAMLKey(w, header, val, termWidth)
+		if err != nil {
+			return err
+		}
+	}
+
+	// --verbose means output all of the fields
+	if x.Verbose {
+		allHeadersMap := assertion.Headers()
+		allHeaders := make([]string, len(allHeadersMap))
+		i := 0
+		for k := range allHeadersMap {
+			allHeaders[i] = k
+			i++
+		}
+
+		sort.Strings(allHeaders)
+		for _, headerName := range allHeaders {
+			header := allHeadersMap[headerName]
+			// don't print any meta headers except revision, which is important
+			// for a device which has remodeled
+			if headerName != "revision" && stringInList(headerName, asserts.MetaHeaders) {
+				continue
+			}
+			if stringInList(headerName, primaryHeaderKeys) {
+				continue
+			}
+
+			switch v := header.(type) {
+			case int, uint, uint64, int64, float32, float64, string, bool:
```

Comment:
note: assertions use strings as their only scalar type.  See for example

https://github.com/snapcore/snapd/blob/master/asserts/headers.go#L274